banyand/liaison/grpc/grpc.go

```diff
@@ -128,3 +169,87 @@ func (s *Server) GracefulStop() {
 	s.log.Info().Msg("stopping")
 	s.ser.GracefulStop()
 }
+
+func assemblyWriteData(shardID uint, writeEntity *v1.WriteRequest, seriesID uint64) data.TraceWriteDate {
+	return data.TraceWriteDate{ShardID: shardID, SeriesID: seriesID, WriteRequest: writeEntity}
+}
+
+func (s *Server) Write(TraceWriteServer v1.TraceService_WriteServer) error {
+	for {
+		writeEntity, err := TraceWriteServer.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		//log.Println("writeEntity:", writeEntity)
+		ana := logical.DefaultAnalyzer()
+		metadata := common.Metadata{
+			KindVersion: apischema.SeriesKindVersion,
+			Spec:        writeEntity.GetMetadata(),
+		}
+		schema, ruleError := ana.BuildTraceSchema(context.TODO(), metadata)
+		if ruleError != nil {
+			return ruleError
+		}
+		if s.seriesInfo.seriesEvent == nil {
+			return errors.New("No seriesEvents")
+		}
+		seriesIDLen := len(s.seriesInfo.seriesEvent.FieldNamesCompositeSeriesId)
+		var str string
+		var arr []string
+		for i := 0; i < seriesIDLen; i++ {
+			id := s.seriesInfo.seriesEvent.FieldNamesCompositeSeriesId[i]
+			if defined, sub := schema.FieldSubscript(id); defined {
+				for _, field := range writeEntity.GetEntity().GetFields() {
+					switch v := field.GetValueType().(type) {
+					case *v1.Field_StrArray:
+						for j := 0; j < len(v.StrArray.Value); j++ {
+							if sub == j {
+								arr = append(arr, v.StrArray.Value[j])
+							}
+						}
+					case *v1.Field_IntArray:
+						for t := 0; t < len(v.IntArray.Value); t++ {
+							arr = append(arr, fmt.Sprint(v.IntArray.Value[t]))
+						}
+					case *v1.Field_Int:
+						arr = append(arr, fmt.Sprint(v.Int.Value))
+					case *v1.Field_Str:
+						arr = append(arr, fmt.Sprint(v.Str.Value))
+					}
+				}
+			}
+		}
+		str = strings.Join(arr, "")
+		if str == "" {
+			return errors.New("invalid seriesID")
+		}
+		seriesID := []byte(str)
+		shardNum := s.shardInfo.shardEvent.GetShard().GetId()
```

Comment:
shardEvent might be nil, pls check whether it's nil before read it. and `sync.RWMutex` is also needed.